internal/proxy/meta_cache.go

```diff
@@ -80,23 +81,46 @@ type collectionInfo struct {
 	collID              typeutil.UniqueID
 	schema              *schemapb.CollectionSchema
 	partInfo            map[string]*partitionInfo
-	shardLeaders        map[string][]nodeInfo
-	leaderMutex         sync.Mutex
+	leaderMutex         sync.RWMutex
+	shardLeaders        *shardLeaders
 	createdTimestamp    uint64
 	createdUtcTimestamp uint64
 	isLoaded            bool
 }
 
-// CloneShardLeaders returns a copy of shard leaders
-// leaderMutex shall be accuired before invoking this method
-func (c *collectionInfo) CloneShardLeaders() map[string][]nodeInfo {
-	m := make(map[string][]nodeInfo)
-	for channel, leaders := range c.shardLeaders {
-		l := make([]nodeInfo, len(leaders))
-		copy(l, leaders)
-		m[channel] = l
+// shardLeaders wraps shard leader mapping for iteration.
+type shardLeaders struct {
+	idx *atomic.Int64
+
+	shardLeaders map[string][]nodeInfo
+}
+
+type shardLeadersIterator struct {
```

Comment:
named as Reader is better, it's behavior not like a iterator

Reply:
fixed